### ISSUE TITLE
[RFC][WebProfilerBundle] Slow kernel.terminate can prevent Toolbar from loading

### ISSUE THREAD

**Issue by JoeKre** [symfony/symfony#25849]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes
| Symfony version  |  >= 3.4

I would like to get your opinion concerning the `WebDebugToolbar` and potentially long running `kernel.terminate` events when sending spooled mails which "breaks" the toolbar

The scenario can be described as follows:

When using the `SwiftMailerBundle` with any type of spooling active i.e. `spool:     { type: memory }` the actual sending will be done through `Symfony\Bundle\SwiftmailerBundle\EventListener\EmailSenderListener` before the kernel is actually terminated. 

This will lead to the `EmailSenderListener` blocking the kernel termination until the response from your email provider is fully processed. 
Until the kernel is fully terminated the profile for the given token cannot be loaded. 
While this is happening the toolbar might already try to load the profile data through its ajax call.

Based on the performance of the email provider or any potential `SwiftMailer plugins` it can happen that the `WebDebugToolbar` will quit after the 5 tries to load the profile and show the error instead. 

Two possible fixes for this problem could be to either: 

- increase the number of retries for the ajax call (possibly through a configurable value)

or

- add a small delay before sending the `404` if the profile could not be loaded and create more time for `kernel.terminate` to finish running.

Simplified example in `Symfony\Bundle\WebProfilerBundle\Controller\ProfilerController::toolbarAction`

```
if (!$profile = $this->profiler->loadProfile($token)) {
    sleep(1);
    return new Response('', 404, array('Content-Type' => 'text/html'));
}

```

Is there any need to fix this "problem" which is mildly annoying although it's not actually a bug in the profiler? 
If someone should fix it, or rather give the toolbar more leeway to load the profile data, which way would be a preferred one?  

kind regards
Joe

**Comment by nicolas-grekas** [comment]
I'm not sure there is anything we should do on Symfony's side. Making background tasks on kernel.terminate should be replaced by a proper job queue when this situation is reached instead.

**Comment by JoeKre** [comment]
In general i agree with you however in this case we are not really talking about making some big tasks on kernel.terminate.
It's more like a combination of use the MailerBundle with spooling enabled and sending a single mail and a slow mail provider slowing down the sending of spooled mails. 

Moving the email sending to a job or setup an infrastructure using exim or anything like that would of course work around the problem but seems like a bit too much for a simple workaround.

**Comment by stof** [comment]
Well, there is 3 options for that IMO:

- keeping the current behavior
- stop profiling the work happening during `kernel.terminate`, allowing to save the profile earlier (which basically reverts to the behavior of earlier 2.x versions of Symfony)
- add more delay before retrying the loading of the toolbar in case of 404, to give it more chance to succeed after `kernel.terminate` is done.

As we already retry the loading of the toolbar, the third solution might be a viable option, by only delaying retries (so that the loading is not delayed too much for most pages not using `kernel.terminate`)

**Comment by stof** [comment]
note that delaying the retry should not be done by blocking the server during 1s before sending the 404 for the previous try. It should rather be done by making the JS wait (in a non-blocking way) before the retry.

**Comment by JoeKre** [comment]
@stof 

guess that could be easily done by small changes to the js without any side effects: 

`toolbar_js.html.twig@124`:
``
{ maxTries: 5, timeoutDelay: 1000 }
``

and `base_js.html.twig@30+`:
```
options.maxTries = options.maxTries || 0;
options.timeoutDelay = options.timeoutDelay || 500;
xhr.open(options.method || 'GET', url, true);
xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
xhr.onreadystatechange = function(state) {
    if (4 !== xhr.readyState) {
        return null;
    }

    if (xhr.status == 404 && options.maxTries > 1) {
        setTimeout(function(){
            options.maxTries--;
            request(url, onSuccess, onError, payload, options);
         }, options.timeoutDelay);

         return null;
    }
```

**Comment by nicolas-grekas** [comment]
@JoeKre would you mind sending a PR?

**Comment by JoeKre** [comment]
First time for everything i guess :)

I hope i didn't completely screw something up with the process for contributing.

**Comment by javiereguiluz** [comment]
Fixed by #25967.
